src/main/kotlin/org/treeWare/proto3/message/EncodeProto3.kt

```diff
@@ -0,0 +1,214 @@
+import org.treeWare.model.core.getMetaModelResolved
+import org.treeWare.model.traversal.TraversalAction
+import org.treeWare.proto3.aux.getProto3MetaModelMap
+import java.io.File
+
+
+fun encodeProto3(mainMetaModel: MainModel, writePath: String) {
+    val protoVisitor = ProtoEncoderVisitor(writePath)
+    metaModelForEach(mainMetaModel, protoVisitor)
+}
+
+
+private class ProtoEncoderVisitor(val writePath: String) : Leader1MetaModelVisitor<TraversalAction> {
+
+    var bodyWriter = PicoWriter()
+    var headerWriter = PicoWriter()
+    var currentPackage = ""
+
+    val keyOptions = mutableListOf<String>()
+    val requiredOptions = mutableListOf<String>()
+    val imports = mutableListOf<String>()
+
+    override fun visitMainMeta(leaderMainMeta1: MainModel): TraversalAction {
+        visitImports(leaderMainMeta1)
+        visitKeyFieldOptions(leaderMainMeta1)
+        visitRequiredFieldOptions(leaderMainMeta1)
+
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveMainMeta(leaderMainMeta1: MainModel) {
+    }
+
+    override fun visitRootMeta(leaderRootMeta1: EntityModel): TraversalAction {
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveRootMeta(leaderRootMeta1: EntityModel) {
+    }
+
+    override fun visitPackageMeta(leaderPackageMeta1: EntityModel): TraversalAction {
+        writeImports()
+        val name = getMetaName(leaderPackageMeta1)
+        currentPackage = name
+        headerWriter.writeln("syntax = \"proto3\";")
+        headerWriter.writeln("")
+        headerWriter.writeln("package $name;")
+        headerWriter.writeln("")
+
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leavePackageMeta(leaderPackageMeta1: EntityModel) {
+        val name = getMetaName(leaderPackageMeta1)
+        val fullPath = "$writePath/$name.proto"
+        val outWriter = File(fullPath).bufferedWriter()
+        outWriter.write(headerWriter.toString())
+        outWriter.write(bodyWriter.toString())
+        outWriter.close()
+        bodyWriter = PicoWriter()
+        headerWriter = PicoWriter()
+    }
+
+    override fun visitEnumerationMeta(leaderEnumerationMeta1: EntityModel): TraversalAction {
+        val enumName = getMetaName(leaderEnumerationMeta1)
+        bodyWriter.writeln_r("enum ${snakeToCamel(enumName)} {")
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveEnumerationMeta(leaderEnumerationMeta1: EntityModel) {
+        bodyWriter.writeln_l("}")
+        bodyWriter.writeln("")
+    }
+
+    override fun visitEnumerationValueMeta(leaderEnumerationValueMeta1: EntityModel): TraversalAction {
+        val enumValue = getMetaName(leaderEnumerationValueMeta1).uppercase()
+        val enumNumber = getMetaNumber(leaderEnumerationValueMeta1)
+        bodyWriter.writeln("$enumValue = $enumNumber")
+        return TraversalAction.CONTINUE
+    }
+
+    override fun leaveEnumerationValueMeta(leaderEnumerationValueMeta1: EntityModel) {
+
```

Comment:
nit: remove blank line. Also, empty function bodies can be on the same line as the function:

`fun someFun() {}`